### PR TITLE
Vet clinic database: add "join table" for visits

### PR DESCRIPTION
**Pull request by nilab-khwaja**
Create a table named vets with the following columns:

- id: integer (set it as autoincremented PRIMARY KEY)
- name: string
- age: integer
- date_of_graduation: date

Create a "join table" called specializations.
Create a "join table" called visits'
Insert the data for visits and other tables .
Write queries to answer the following:

- Who was the last animal seen by William Tatcher?
- How many different animals did Stephanie Mendez see?
- List all vets and their specialties, including vets with no specialties.
- List all animals that visited Stephanie Mendez between April 1st and August 30th, 2020.
- What animal has the most visits to vets?
- Who was Maisy Smith's first visit?
- Details for most recent visit: animal information, vet information, and date of visit.
- How many visits were with a vet that did not specialize in that animal's species?
- What specialty should Maisy Smith consider getting? Look for the species she gets the most.
q1-2
q3

q4
q5
q6
q7
q8
q9

### CODE REVIEW

File: data.sql [hunk 1]
```diff
@@ -5,4 +5,27 @@ INSERT INTO animals (name,date_of_birth, escape_attempts, neutered, weight_kg) V
...
+UPDATE animals SET owner_id = (SELECT id FROM owners WHERE full_name = 'Bob') WHERE name IN ('Devimon', 'Plantmon');
+
+UPDATE animals SET owner_id = (SELECT id FROM owners WHERE full_name = 'Melody Pond') WHERE name IN ('Charmander', 'Squirtle', 'Blossom');
+
+UPDATE animals SET owner_id = (SELECT id FROM owners WHERE full_name = 'Dean Winchester') WHERE name IN ('Angemon', 'Boarmon');
+
+
+INSERT INTO vets(name,age,date_of_graduation) VALUES ('Vet William Tatcher', 45, '2000-04-23'), ('Vet Maisy Smith', 26, '2019-01-17'), ('Vet Stephanie Mendez',64, '1981-05-04'), ('Jack Harkness', 38, '2008-06-08');
+
+INSERT INTO visits (animal_id, vet_id , visit_date) VALUES (1,1,'2020-05-24'), (1,3, '2020-07-22'), (2,4,'2021-02-02'), (3,2,'2020-01-05'), (3,2,'2020-03-08'), (3,2,'2020-05-14'), (4,3,'2021-05-04'), (6,2,'2019-12-21'), (6,1,'2020-08-10'),(6,2, '2021-04-07'), (7,3,'2019-09-29'), (8,4,'2020-10-03'), (8,4,'2020-11-04'), (9,2,'2019-01-24'), (9,2,'2019-05-15'),(9,2,'2020-02-20'), (9,2,'2020-08-03'), (10,3,'2020-05-24'), (10,1,'2021-01-11');
```
Comment (mckent05) [inline data.sql:31]: - Good job done so far, but I cannot see the data you added into the specializations table, please note it is important to have all the added data in your `data.sql` file. Please fix this 🙏

File: queries.sql [hunk 2]
```diff
@@ -43,3 +43,24 @@ SELECT neutered, COUNT(*) AS escape_count  FROM animals WHERE escape_attempts >
...
+SELECT a.name AS animal_name , COUNT(v.animal_id) AS visit_count from animals a JOIN visits v ON v.animal_id = a.id  GROUP BY a.name  ORDER BY visit_count DESC LIMIT 1;
+SELECT vt.name AS Vet_name FROM visits v JOIN vets vt ON v.vet_id = vt.id JOIN animals a ON v.animal_id = a.id WHERE a.name = 'Vet Maisy Smith' ORDER BY v.visit_date ASC LIMIT 1;
+SELECT a.name AS animal_name  FROM visits v JOIN vets vt ON v.vet_id = vt.id JOIN animals a ON v.animal_id = a.id WHERE a.name = 'Vet Maisy Smith' ORDER BY v.visit_date ASC LIMIT 1;
+SELECT vt.name AS Vet_name FROM visits v JOIN vets vt ON v.vet_id = vt.id JOIN animals a ON v.animal_id = a.id WHERE a.name = 'Vet Maisy Smith' ORDER BY v.visit_date ASC LIMIT 1;
+SELECT a.name  AS animal_name , vt.name AS vet_name, v.visit_date FROM visits v JOIN vets vt ON v.vet_id = vt.id JOIN animals a ON v.animal_id = a.id ORDER BY v.visit_date  DESC LIMIT 1;
+SELECT COUNT(*) AS "Visits Without Specialization"FROM visits v JOIN animals a ON v.animal_id = a.id JOIN vets vet ON v.vet_id = vet.id LEFT JOIN specializations s ON vet.id = s.vet_id AND a.species = (SELECT species FROM species WHERE id = s.species_id) WHERE s.vet_id IS NULL;
```
Comment (mckent05) [inline queries.sql]: **OPTIONAL** - For clarity's sake, it'd be a lot better if you state the question each of your queries is to answer just above the query. It'd help anyone viewing your project understand what you are trying to achieve. Please fix this 😄 
- For this query `How many different animals did Stephanie Mendez see?`, I could not find the query or the answer to the query in your PR, please include the query and attach the answer in your PR 🙏 
- For this query `Who was Maisy Smith's first visit?`, I could not find the query or the answer to the query in your PR, please include the query and attach the answer in your PR 🙏

File: queries.sql [hunk 3]
```diff
@@ -43,3 +43,24 @@ SELECT neutered, COUNT(*) AS escape_count  FROM animals WHERE escape_attempts >
...
+SELECT a.name FROM animals a JOIN species s ON a.species_id = s.id WHERE s.name = 'Pokemon';
+SELECT o.full_name , a.name AS animal_name FROM owners o JOIN animals a ON a.owner_id = o.id;
+SELECT s.name, COUNT(a.id) AS animal_count FROM species s LEFT JOIN animals a ON s.id = a.species_id GROUP BY s.name;
+SELECT o.full_name ,COUNT(a.id) AS animal_count FROM owners o LEFT JOIN animals a ON a.owner_id = o.id GROUP BY o.full_name ORDER BY animal_count DESC LIMIT 1;
+SELECT a.name FROM animals a JOIN owners o ON a.owner_id = o.id WHERE o.full_name = ' Dean Winchester' AND a.escape_attempts = 0;
+SELECT a.name FROM animals a JOIN owners o ON a.owner_id = o.id JOIN species s  ON a.species_id = s.id WHERE  o.full_name = 'Jennifer Orwell' AND s.name = 'Digimon';
+
+
+
+SELECT vt.name  AS vet_name from vets vt LEFT JOIN specializations sp ON vt.id = sp.vet_id LEFT JOIN species s ON sp.species_id = s.id;
```
Comment (mckent05) [inline queries.sql]: - You have done well so far, but the answer to this query is wrong, you are required to list the last animal seen by William Tatcher. Your answer should be 'Blossom'. Please fix this and attach the correct answer in your PR.
  ```SQL
       SELECT animals.name, vistis.visit_date
        FROM visits
       LEFT JOIN animals ON animals.id = visits.animal_id
       LEFT JOIN vets ON vets.id = visits.vet_id
       WHERE vets.name = 'William Tatcher'
      ORDER BY visits.visit_date DESC
       LIMIT 1;
```

File: queries.sql (hunk 2 above)
Comment (mckent05) [inline queries.sql]: - You have done well so far, but the attached answer in your PR, for this query, `How many times visits where with a vet that did not specialize in that animal's species?`, is wrong. Please re-run your query and attach the correct answer in your PR. 
``` SQL
       SELECT count(*)
      FROM visits
      LEFT JOIN animals ON animals.id = visits.animal_id
      LEFT JOIN vets ON vets.id = visits.vet_id
      WHERE animals.species_id NOT IN (SELECT species_id FROM specializations WHERE vet_id = vets.id)
```

File: queries.sql [hunk 4]
```diff
@@ -43,3 +43,24 @@ SELECT neutered, COUNT(*) AS escape_count  FROM animals WHERE escape_attempts >
...
+SELECT a.name AS animal_name  FROM visits v JOIN vets vt ON v.vet_id = vt.id JOIN animals a ON v.animal_id = a.id WHERE a.name = 'Vet Maisy Smith' ORDER BY v.visit_date ASC LIMIT 1;
+SELECT vt.name AS Vet_name FROM visits v JOIN vets vt ON v.vet_id = vt.id JOIN animals a ON v.animal_id = a.id WHERE a.name = 'Vet Maisy Smith' ORDER BY v.visit_date ASC LIMIT 1;
+SELECT a.name  AS animal_name , vt.name AS vet_name, v.visit_date FROM visits v JOIN vets vt ON v.vet_id = vt.id JOIN animals a ON v.animal_id = a.id ORDER BY v.visit_date  DESC LIMIT 1;
+SELECT COUNT(*) AS "Visits Without Specialization"FROM visits v JOIN animals a ON v.animal_id = a.id JOIN vets vet ON v.vet_id = vet.id LEFT JOIN specializations s ON vet.id = s.vet_id AND a.species = (SELECT species FROM species WHERE id = s.species_id) WHERE s.vet_id IS NULL;
+SELECT s.species FROM visits v JOIN animals a ON v.animal_id = a.id JOIN vets vet ON v.vet_id = vet.id JOIN specializations sp ON vet.id = sp.vet_id JOIN species s ON sp.species_id = s.id WHERE a.name = 'Vet Maisy Smith' GROUP BY s.species ORDER BY COUNT(*) DESC LIMIT 1;
```
Comment (mckent05) [inline queries.sql]: - You have done well so far, but the attached answer in your PR, for this query, `What specialty should Maisy Smith consider getting? Look for the species she gets the most`, is wrong. Please re-run your query and attach the correct answer in your PR.
```sql
        SELECT species.name, count(*)
    FROM visits
    LEFT JOIN animals ON animals.id = visits.animal_id
    LEFT JOIN species ON animals.species_id = species.id
    LEFT JOIN vets ON vets.id = visits.vet_id
    WHERE vet.name = 'Maisy Smith'
    GROUP BY species.name
```

**Review by mckent05 — CHANGES_REQUESTED** [review]
Hi @nilab-khwaja, 

# STATUS: CHANGES REQUIRED ♻️ ♻️ 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!
almost there

## Highlights
- Good job adding join tables for visits ✔️ ✔️ 
- Your PR is professional 👍 👍 


## Required Changes ♻️
- You have done well so far, but please take your time to input the correct data into your database and ensure you get all queries and their answers correctly 👍 👍 
_Check the comments under the review._

## Optional suggestions
Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better.

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by thecodechaser — APPROVED** [review]
## STATUS: APPROVED 🟢

## Hi @nilab-khwaja👋, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## To Highlight 🎉
- Good job with join tables. :heavy_check_mark:
- Gitflow is followed. :heavy_check_mark:
- Professional README file. :heavy_check_mark:
- PR has a good title and summary. ✔️



Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**